File: leetcode/weekly-contest/20240317/100227-拾起K个1需要的最少行动次数.py
from cmath import inf
from itertools import accumulate
from typing import List

# class Solution:
#     def minimumMoves(self, nums: List[int], k: int, maxChanges: int) -> int:
#         acc = list(accumulate(nums, initial=0))
#         n = len(nums)
#         pos_sum = list(accumulate((i if nums[i] else 0 for i in range(n)), initial=0))
        
#         result = inf
#         for i in range(n):
#             ans = 0
#             tmp = k

#             # 将 i - 1, i, i + 1 的 1 用掉
#             for p in range(i - 1, i + 2):
#                 if 0 <= p < n and nums[p] and tmp:
#                     tmp -= 1
#                     ans += abs(p - i) # 在 i - 1, i + 1 时要进行移动操作

#             if maxChanges >= tmp: # maxChanges 够用
#                 ans += tmp * 2    # 赋值 tmp 次在旁边，然后移动到 index 有 tmp 次，总共 2 * tmp 次
#             else:                 # 把临近的和 maxChanges 都用掉还不够
#                 ans = 2 * maxChanges
#                 tmp = k - maxChanges

#                 l, r = 0, n
#                 while l <= r:
#                     m = (l + r) // 2
#                     # 以 i 为中心 m 为半径，正好有 tmp 以上的 1
#                     if acc[min(n, i + m + 1)] - acc[max(0, i - m)] >= tmp:
#                         r = m - 1
#                     else:
#                         l = m + 1
#                 m = l

#                 # i 右侧的 1
#                 ans += (pos_sum[min(n, i+m+1)] - pos_sum[i]) - (acc[min(n, i+m+1)] - acc[i]) * i
#                 # i 左侧的 1
#                 ans += -(pos_sum[i+1] - pos_sum[max(0, i-m)]) + (acc[i+1] - acc[max(0, i-m)]) * i

#                 # 注意这样选取区间，区间元素个数可能比需要的多一个，那个元素需要去除
#                 if acc[min(n, i + m + 1)] - acc[max(0, i - m)] > tmp:
#                     ans -= m

#             # 更新 result
#             result = min(ans, result)

#         return result


class Solution:
    def minimumMoves(self, nums: List[int], k: int, maxChanges: int) -> int:
        pos = []
        c = 0
        n = len(nums)
        for i in range(n):
            if nums[i]:
                pos.append(i)
                c = max(c, 1)
                if i + 1 < n and nums[i + 1]:
                    if i + 2 < n and nums[i + 2]:
                        c = 3
                    else:
                        c = max(c, 2)

        c = min(k, c) # k - c >= 0
        if maxChanges >= k - c:
            # c = 0 -> 0, c = 1 -> 0, c = 2 -> 1, c = 3 -> 2
            return max(c - 1, 0) + (k - c) * 2

        ans = inf
        n = len(pos)
        size = k - maxChanges
        prefix = list(accumulate(pos, initial=0))

        for right in range(size, n + 1):
            #  s1+s2 是 j 在 [left, right) 中的所有 pos[j] 到 pos[(left+right)/2] 的距离之和
            left = right - size
            mid = left + size // 2
            tmp = pos[mid] * (mid - left) - (prefix[mid] - prefix[left])
            tmp += prefix[right] - prefix[mid] - pos[mid] * (right - mid)
            ans = min(ans, tmp)

        return ans + 2 * maxChanges

                        

File: problem-sets/leetcode-weekly-contest/20240317/100227-拾起K个1需要的最少行动次数.cpp
#include <algorithm>
#include <limits>
#include <vector>

using std::vector;

class Solution
{
public:
  long long minimumMoves(vector<int> &nums, int k, int maxChanges)
  {
    int c = 0;
    vector<int> pos;
    for (int i = 0; i < nums.size(); ++i) {
      if (nums[i] == 1) {
        pos.emplace_back(i);
        c = std::max(c, 1);
        if (i + 1 < nums.size() && nums[i + 1] == 1) {
          if (i + 2 < nums.size() && nums[i + 2] == 1) {
            c = 3;
          } else {
            c = std::max(c, 2);
          }
        }
      }
    }

    c = std::min(c, k);
    if (maxChanges >= k - c) {
      return std::max(c - 1, 0) + 2 * (k - c);
    }

    int n = pos.size();
    vector<long long> prefix(n + 1);
    for (int i = 1; i <= n; ++i) {
      prefix[i] = prefix[i - 1] + pos[i - 1];
    }

    int size = k - maxChanges;
    long long ans = std::numeric_limits<long long>::max();

    for (int right = size; right <= n; ++right) {
      int left = right - size;
      int mid = left + (size >> 1);
      long long midPos = pos[mid];
      long long s1 = midPos * (mid - left) - (prefix[mid] - prefix[left]);
      long long s2 = (prefix[right] - prefix[mid]) - midPos * (right - mid);
      ans = std::min(ans, s1 + s2);
    }
    return ans + 2 * maxChanges;
  }
};